src/worker/indexers/chainIndexer.ts

```diff
@@ -0,0 +1,285 @@
+    service: "worker",
+    level: "debug",
+    message: `Finished getting contracts - ChainIndexer: ${
+      params.chainId
+    }, Time taken: ${(endContractsTime - startContractsTime).toFixed(2)} ms`,
+  });
+
+  // cache the blocks and their timestamps
+  const startBlocksTime = performance.now();
+  logger({
+    service: "worker",
+    level: "debug",
+    message: `Starting to get blocks - ChainIndexer: ${params.chainId}`,
+  });
+  const uniqueBlockNumbers = [
+    ...new Set<number>(logs.map((log) => log.blockNumber)),
+  ];
+  const blockDetails = await Promise.all(
+    uniqueBlockNumbers.map(async (blockNumber) => ({
+      blockNumber,
+      details: await provider.getBlock(blockNumber),
+    })),
+  );
+  const blockCache = blockDetails.reduce((acc, { blockNumber, details }) => {
+    acc[blockNumber] = details;
+    return acc;
+  }, {} as Record<number, ethers.providers.Block>);
+  const endBlocksTime = performance.now();
+  logger({
+    service: "worker",
+    level: "debug",
+    message: `Finished getting blocks - ChainIndexer: ${
+      params.chainId
+    }, Time taken: ${(endBlocksTime - startBlocksTime).toFixed(2)} ms`,
+  });
+
+  // format the logs to ContractLogEntries
+  const formattedLogs = logs.map((log) => {
+    const contractAddress = log.address;
+
+    // attempt to decode the log
+    let decodedLog;
+    let decodedEventName;
+
+    const contract = contractCache[contractAddress];
+    if (contract) {
+      try {
+        const iface = new ethers.utils.Interface(contract.abi);
+        const parsedLog = iface.parseLog(log);
+        decodedEventName = parsedLog.name;
+        decodedLog = parsedLog.eventFragment.inputs.reduce((acc, input) => {
+          acc[input.name] = {
+            type: input.type,
+            value: parsedLog.args[input.name].toString(),
+          };
+          return acc;
+        }, {} as Record<string, { type: string; value: string }>);
+      } catch (error) {
+        logger({
+          service: "worker",
+          level: "warn",
+          message: `Failed to decode log: chainId: ${params.chainId}, contractAddress ${contractAddress}`,
+        });
+      }
+    }
+
+    const block = blockCache[log.blockNumber];
+
+    // format the log entry
+    return {
+      chainId: params.chainId,
+      blockNumber: log.blockNumber,
+      contractAddress: log.address.toLowerCase(), // ensure common address handling across
+      transactionHash: log.transactionHash,
+      topic0: log.topics[0],
+      topic1: log.topics[1],
+      topic2: log.topics[2],
+      topic3: log.topics[3],
+      data: log.data,
+      eventName: decodedEventName,
+      decodedLog: decodedLog,
+      timestamp: new Date(block.timestamp * 1000), // ethers timestamp is s, Date uses ms
+      transactionIndex: log.transactionIndex,
+      logIndex: log.logIndex,
+    } as ContractEventLogEntry;
+  });
+
+  return formattedLogs;
+};
+
+export const createChainIndexerTask = async (chainId: number) => {
```

Comment:
can we move this into the `tasks` folder with a new file. to be consistent on the structure of files/folders